### PR TITLE
[hybrid performance] optim the grad fuse for pipeline mode by sorting the grad by dtype

### PR DESCRIPTION
**Pull request by FeixLiu**
### PR types
Performance optimization

### PR changes
Others

### Describe
Sorted the grad by dtype before coalescing, can decrease the number of coalescing op.
Besides, by reducing the number of coalesce op, the number of c_allreduce_sum op can also be reduced.

_All test are based on Ernie3.0, pp=dp=mp=2_
_fp16_allreduce=True_ 
_optimize_cast=True_ 

#### Decrease in number of coalesce op
|| number before this opt | number after this opt | gain | 
|:--:|:--:|:--:|:--:|
|card 0/4| 48| 6 | -700% |
|card 1/5| 48|6| -700%|
|card 2/6| 120|12 | -900%|
|card 3/7| 120| 12| -900%|
|one dp group|336 | 36| -833%|

#### Loss diff
Screen Shot 2021-08-23 at 10 44 10 AM
Screen Shot 2021-08-23 at 10 44 24 AM

#### Performance 
| throughput  before this opt | throughput after this opt | gain | 
|:--:|:--:|:--:|
| 39146 | 39232| + 0.2%|

### CODE REVIEW

**Review by wangxicoding — APPROVED** [review]
LGTM。 性能再给一下

File: python/paddle/fluid/optimizer.py [hunk 1]
```diff
@@ -5409,6 +5412,24 @@ def _accumulate_gradients_with_fuse(self, main_block, fp16, fused_size):
...
+        other_pairs = []
+        for pairs in grad_param_pairs:
+            dtype = main_block.var(pairs[0]).dtype
+            if dtype == paddle.float32:
+                fp32_pairs.append(pairs)
+            elif dtype == paddle.float16:
+                fp16_pairs.append(pairs)
+            else:
+                other_pairs.append(pairs)
+        sorted_pairs = fp16_pairs
```
Comment (gongweibao) [inline python/paddle/fluid/optimizer.py:5428]: 这个已经打乱了grad生成顺序了。重新改一下这个函数，保证grad的顺序在一定范围内打乱次序。

**Review by gongweibao — CHANGES_REQUESTED** [review]
看comment。

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**